### ISSUE TITLE
ENH Add post-processing algorithm "Discrimination Aware Decision Tree Learning" by Kamiran et al.

### ISSUE THREAD

**Issue by hildeweerts** [fairlearn/fairlearn#1024]
The paper [Discrimination Aware Decision Tree Learning](https://www.win.tue.nl/~mpechen/publications/pubs/KamiranICDM2010.pdf) by Kamiran et al. introduces several approaches for learning a decision tree that satisfies demographic parity. Given that the relabeling procedure clearly outperforms the other approaches, the goal of this task is to implement the **relabeling** procedure to Fairlearn such that it is compatible with [sklearn.tree.DecisionTreeClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn-tree-decisiontreeclassifier).

Completing this item requires:
- [ ] postprocessing technique code in `fairlearn.postprocessing`
- [ ] unit tests in `test.unit.postprocessing`
- [ ] descriptive API reference (directly in the docstring)
- [ ] a short user guide in `docs.user_guide.mitigation.rst` 

A fully fledged example notebook is not required.

To claim this task please respond below. Of course, you can also ask questions!

**Comment by romanlutz** [comment]
@Hadyark do you mean for Fairlearn? @hildeweerts tagged it as help wanted so you should feel free to draft a PR. I would recommend publishing it early on to get feedback rather than spending a lot of time and then having to revise later on, but that's just me...
Please upvote if you want to get started and we can assign it go you! 🙂

**Comment by Hadyark** [comment]
@romanlutz Yes, for Fairlearn but do I have to relabel a decision tree from Sklearn? Or maybe I misunderstood and you have your own implementation of a decision tree.

Because right now I can relabel a decision tree from our implementation but I have to adapt my code to relabel a decision tree from Sklearn (because the structure of the tree is different, see [here](https://scikit-learn.org/stable/auto_examples/tree/plot_unveil_tree_structure.html#sphx-glr-auto-examples-tree-plot-unveil-tree-structure-py))

**Comment by romanlutz** [comment]
No, we don't have a custom implementation. What you wrote matches my expectations, just wanted to clarify to make sure we are thinking of the same thing. 🙂

**Comment by Hadyark** [comment]
I have a problem. It seems that it is not possible to relabel a leaf in a tree made by Sklearn. I tried to modify the clf.tree_ variable but it doesn't change the predictions. So I think this is something that is used in Cython and I think it is not possible to influence this part

**Comment by Hadyark** [comment]
I know how to relabel a leaf but I need to modify the "clf.tree_.value". This is an array that contains the number of samples for each class and is used for prediction. For example, if I have for a leaf the array [10., 5.]. I have 10 elements for class 0 and 5 for class 1. and the leaf will predict the class 0 (the majority class). So for relabel, I reverse the elements and I get [5., 10.].

 I don't know if this has any consequences for you.

**Comment by romanlutz** [comment]
I don't see a problem with that, but I also haven't worked with the internals of sklearn's decision trees before. Tagging @fairlearn/fairlearn-maintainers to chime in if they have thoughts.

**Comment by hildeweerts** [comment]
Hi, @Hadyark! Relabeling by swapping elements seems reasonable to me. A possible caveat is that the meaning of the proportions in the leafs is lost after swapping, so we might want to disable `predict_proba()` and `predict_log_proba()` (i.e., only keep `predict()`).

On a related note, `sklearn.tree.DecisionTreeClassifier` is much more flexible than the algorithm proposed in the paper by Kamiran et al. For example, the paper assumes binary classification, whereas the scikit-learn implementation allows for multiclass classification. So we should ensure that only the relevant functionality of sklearn's decision tree is included in the fairness-aware decision tree.

**Comment by adrinjalali** [comment]
Note that if this implementation requires changing the `clf.tree_` code, it means vendoring and changing Cython code and adding Cython as a dependency and add it to the build process. The wheel build process would be significantly more complicated and we'd need to release wheels for every single platform and compile for them separately.

But if we can use a `DecisionTree*` and just modify the values in `tree_`, then it should be much easier.

Also, note that in practice on large datasets, `DecisionTree*` based models are really slow and not recommended for use in the industry, and people should be using the quantized `HistGradientBoosting*` models instead. I haven't read the paper, and I'm not sure if one can implement it for the latter trees.

**Comment by Hadyark** [comment]
> On a related note, `sklearn.tree.DecisionTreeClassifier` is much more flexible than the algorithm proposed in the paper by Kamiran et al. For example, the paper assumes binary classification, whereas the scikit-learn implementation allows for multiclass classification. So we should ensure that only the relevant functionality of sklearn's decision tree is included in the fairness-aware decision tree.

I only handled the case of a binary tree because it was similar to my work. If it is necessary to adapt it for multiclass classification, it will be necessary to think about which class it is better to relabel a leaf and how is calculated the discrimination.

> Note that if this implementation requires changing the `clf.tree_` code, it means vendoring and changing Cython code and adding Cython as a dependency and add it to the build process. The wheel build process would be significantly more complicated and we'd need to release wheels for every single platform and compile for them separately.
> 
> But if we can use a `DecisionTree*` and just modify the values in `tree_`, then it should be much easier.

I just modify the values in `tree_`

**Comment by hildeweerts** [comment]
> I only handled the case of a binary tree because it was similar to my work. If it is necessary to adapt it for multiclass classification, it will be necessary to think about the best class to relabel a leaf and how is calculated the discrimination.

I'm sorry for the confusion - what I meant is that sklearn's decision tree by default accepts multiclass classification problems, whereas the leaf relabeling algorithm as proposed in the paper only works for binary classification. So we should ensure that our implementation (even though it relies on the sklearn decision tree) *only* accepts binary classification problems and throws an error if you do pass a `y` with >2 categories.

> Also, note that in practice on large datasets, `DecisionTree* `based models are really slow and not recommended for use in the industry, and people should be using the quantized `HistGradientBoosting*` models instead. I haven't read the paper, and I'm not sure if one can implement it for the latter trees.

I think it would probably be possible to implement the algorithm for boosted trees, but I'm not sure whether we should really try to do that at this point. For me, the primary goal of adding this algorithm is to use it for benchmarking purposes. My suggestion would be to stay as close to the original paper as possible and leave extensions to other tree-based algorithms for future PRs.

**Comment by Hadyark** [comment]
What should I do for the item "a short user guide in [docs.user_guide.mitigation.rst](https://github.com/fairlearn/fairlearn/blob/main/docs/user_guide/mitigation.rst)" ? 

Should I just fill in the table? Or add a sub-section in the pre-processing? 
If I have to add a sub-section, should I explain what relabelling is or add a mini-tutorial on how to use it?
